pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/parser/RuntimeASTTransformer.groovy

```diff
@@ -932,8 +997,375 @@ class RuntimeASTTransformer {
+            // However, class loading isn't free, so we also don't want a single method per class
+            ClassNode classNode = methodClassNode[groupName]
+            // If we don't have a classNode for this group name or if this class has reached groupSize, start a new class
+            if (classNode == null || classNode.methods.size() >= groupSize) {
+                // Get an uncreative unique class name
+                String className = createStableUniqueName(groupName)
+
+                classNode = new ClassNode("generated." + className, ACC_PUBLIC, ClassHelper.make(RuntimeContainerBase.class))
+                classNode.addConstructor(ACC_PUBLIC, [new Parameter(ClassHelper.make(CpsScript.class), "workflowScript")] as Parameter[], [ClassHelper.make(IOException.class)] as ClassNode[],
+                        block(ctorSuperS(varX("workflowScript"))))
+
+                this.moduleNode.addClass(classNode)
+
+                // Add an instance of this class to the variables that can be referenced from anywhere in this script
+                createScriptContextVariable(ctorX(classNode, args(varX('this'))), className)
+
+                methodClassNode[groupName] = classNode
+            }
+
+            String className = classNode.nameWithoutPackage
+            Expression variable = varX(new DynamicVariable(className, false))
+
+            // Uncreative method naming is fine
+            int methodCount = classNode.methods.size()
+            String name = "get${groupName}_${methodCount}"
+
+            // The instance method referenced via script binding
+            MethodNode method = new MethodNode(name, ACC_PUBLIC, returnType, [] as Parameter[], [] as ClassNode[], methodBody)
+            classNode.addMethod(method)
+
+            return callX(variable, name)
+        }
+
+        /**
+         * Turns a list of expressions into a method that returns a list (not a ListExpression)
+         * Works around limitations on list expression literals
+         * (literal "[item0, item1, ...]") being limited to 250 items.
+         * @param expressions the list expressions to be wrapped in a function call
+         * @return call to a method that returns the provided ListExpression
+         */
+        @Nonnull
+        private Expression toExternalListMethod(@Nonnull Iterable<Expression> expressions) {
+            ListExpression currentListExpression = new ListExpression()
+
+            final int listLimit = 250
+
+            // Local variable in groovy ast is a variable that ... accesses itself.
+            VariableExpression variable = varX(createStableUniqueName('extendedList'))
+            variable.accessedVariable = variable
+
+            Expression declarationExpression = new DeclarationExpression(variable, ASSIGN, new ListExpression())
+            BlockStatement block = block()
+            block.addStatement(stmt(declarationExpression))
+
+            int count = 0
+            expressions.each { item ->
```

Comment:
FWIW I have long advocated that `src/main/groovy/` here be rewritten to Java for maintainability. (`src/main/resources/**/*.groovy` is unrelated—that is CPS-transformed.)

Reply:
I don't understand what that has to do with this line of code.

Reply:
Not this line specifically, just the whole long blocks of new Groovy.